packages/replicache/src/kv/sqlite-store.ts

```diff
@@ -0,0 +1,340 @@
+  finalize(): void;
+}
+
+export interface SQLiteDatabase {
+  /**
+   * Close the database connection.
+   */
+  close(): void;
+
+  /**
+   * Destroy or delete the database (e.g. delete file).
+   */
+  destroy(): void;
+
+  /**
+   * Prepare a SQL string, returning a statement you can execute.
+   * E.g. `const stmt = db.prepare("SELECT * FROM todos WHERE id=?");`
+   */
+  prepare(sql: string): PreparedStatement;
+}
+
+type SQLitePreparedStatements = {
+  beginImmediate: PreparedStatement;
+  commit: PreparedStatement;
+  rollback: PreparedStatement;
+
+  get: PreparedStatement;
+  put: PreparedStatement;
+  del: PreparedStatement;
+};
+
+const getPreparedStatementsForSQLiteDatabase = (
+  db: SQLiteDatabase,
+): SQLitePreparedStatements => ({
+  beginImmediate: db.prepare('BEGIN IMMEDIATE'),
+  commit: db.prepare('COMMIT'),
+  rollback: db.prepare('ROLLBACK'),
+
+  get: db.prepare('SELECT value FROM entry WHERE key = ?'),
+
+  put: db.prepare('INSERT OR REPLACE INTO entry (key, value) VALUES (?, ?)'),
+  del: db.prepare('DELETE FROM entry WHERE key = ?'),
+});
+
+const rwLocks = new Map<string, RWLock>();
+
+/**
+ * A SQLite-based Store implementation.
+ *
+ * This store provides a generic SQLite implementation that can be used with different
+ * SQLite providers (expo-sqlite, better-sqlite3, etc). It implements the Store
+ * interface using a single 'entry' table with key-value pairs.
+ */
+export class SQLiteStore implements Store {
+  readonly #name: string;
+  readonly #dbm: SQLiteDatabaseManager;
+  readonly #rwLock: RWLock;
+
+  readonly #db: SQLiteDatabase;
+
+  readonly #preparedStatements: SQLitePreparedStatements;
+
+  #closed = false;
+
+  constructor(
+    name: string,
+    dbm: SQLiteDatabaseManager,
+    opts?: SQLiteDatabaseManagerOptions | undefined,
+  ) {
+    this.#name = name;
+    this.#dbm = dbm;
+    this.#rwLock = rwLocks.get(name) ?? new RWLock();
+    rwLocks.set(name, this.#rwLock);
+
+    this.#db = this.#dbm.open(name, opts);
+
+    this.#preparedStatements = getPreparedStatementsForSQLiteDatabase(this.#db);
+  }
+
+  async read(): Promise<Read> {
+    const release = await this.#rwLock.read();
+    try {
+      const read = new SQLiteStoreRead(this.#preparedStatements, release);
+
+      return Promise.resolve(read);
+    } catch (e) {
+      return Promise.reject(e);
+    }
+  }
+
+  async write(): Promise<Write> {
+    const release = await this.#rwLock.write();
+    try {
+      const write = new SQLiteStoreWrite(this.#preparedStatements, release);
+
+      return Promise.resolve(write);
+    } catch (e) {
+      return Promise.reject(e);
+    }
+  }
+
+  close(): Promise<void> {
+    for (const stmt of Object.values(this.#preparedStatements)) {
+      stmt.finalize();
+    }
+
+    this.#dbm.close(this.#name);
+    this.#closed = true;
+
+    return promiseVoid;
+  }
+
+  get closed(): boolean {
+    return this.#closed;
+  }
+}
+
+export class SQLiteStoreRead implements Read {
```

Comment:
Curious why the reads are not run in a sqlite transaction.

Reply:
This should now be fixed